src/lib.rs

```diff
@@ -725,16 +733,70 @@ impl<M: ManageConnection> Pool<M> {
         if broken {
             let _ = drop_connections(&inner, locked, vec![conn]).await;
         } else {
-            let conn = IdleConn::make_idle(Conn {
-                conn: conn,
-                birth: birth,
-            });
+            let conn = IdleConn::make_idle(Conn { conn, birth });
             locked.put_idle_conn(conn);
         }
 
         r.map_err(|e| RunError::User(e))
     }
 
+    /// Return connection back in to the pool
+    fn put_back(&self, birth: Instant, mut conn: Conn<M::Connection>) {
+        let inner = self.inner.clone();
+
+        // Supposed to be fast, but do it before locking anyways.
+        let broken = inner.manager.has_broken(&mut conn.conn);
+
+        futures::executor::block_on(async {
+            let mut locked = inner.internals.lock().await;
```

Comment:
It looks like this code is duplicated from `run()`. It should probably be factored out so that `run()` and `put_back()` can share as much logic as possible. (Probably better to move the `broken` value inside the async block, too.)